`std.testing.expectEqual` doesn't compile for `union` with comptime-only field

### Zig Version

0.14.0-dev.39+1b728e183

### Steps to Reproduce and Observed Behavior

Not sure whether this is a regression or it's always been like this,
but it's an unfortunate scenario to have to debug:

```zig
const U = union(enum) {
    a: void,
    b: void,
    c: comptime_int, //remove this field to fix the issue
};

test {
    const expectEqual = @import("std").testing.expectEqual;
    try expectEqual(U{.a = {}}, .a); //comptime before the call fixes the issue
}
```

```
C:\use\Development\zig-windows-x86_64-0.13.0-dev.211+6a65561e3\lib\std\testing.zig:153:37: error: access of union field 'b' while field 'a' is active
                    try expectEqual(@field(expected, fld.name), @field(actual, fld.name));
                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~
main.zig:1:11: note: union declared here
const U = union(enum) {
          ^~~~~
referenced by:
    expectEqual: C:\use\Development\zig-windows-x86_64-0.13.0-dev.211+6a65561e3\lib\std\testing.zig:58:28
    expectEqualInner__anon_1910: C:\use\Development\zig-windows-x86_64-0.13.0-dev.211+6a65561e3\lib\std\testing.zig:153:25
```

It seems that when `U` has a field of comptime-only type, the compiler wants to comptime-evaluate all `@field` expressions.
The code  in `std.testing.expectEqualInner` uses [an `inline for` to access all possible fields](https://github.com/ziglang/zig/blob/04e08ea883f94d2de7a91daee72ccc9613a18a43/lib/std/testing.zig#L151-L156), with the intention of only taking the branch of the active field.
However, all branches are generated (because the `std.meta.eql` call is evaluated at runtime), meaning the `@field` expressions for inactive fields are generated as well.

### Expected Behavior

<details><summary>previous text:</summary>
<p>

I assume the issue is the comptime-evaluation of the `@field`-expression.
A workaround in the `std` implementation may be possible, but afaik this is generally the recommended way of implementing generic union field access.
Having this pattern break with comptime-only types is a bit of a shame.

A language solution would probably be rather complex though:
Either automatically upgrade more of the function evaluation to become `comptime`,
or delay the current compile error to become a runtime panic that triggers only if that code path is reached.

Explicitly marking [the call to `std.mem.eql`](https://github.com/ziglang/zig/blob/04e08ea883f94d2de7a91daee72ccc9613a18a43/lib/std/testing.zig#L152) `comptime` makes the given example work, though it's obviously not feasible for callsites with runtime values.
Maybe there's the option for a middle-ground where marking the call .always_inline should allow the branching to be resolved at comptime for arguments of comptime-only types?

</p>
</details> 

EDIT: Actually scrap all that, we can compare the union tag instead of using std.mem.eql,
which is already comptime-evaluated when possible, fixing the issue.
PR incoming.

--- a/lib/std/meta.zig
+++ b/lib/std/meta.zig
@@ -757,16 +757,13 @@ pub fn eql(a: anytype, b: @TypeOf(a)) bool {
         },
         .Union => |info| {
             if (info.tag_type) |UnionTag| {
-                const tag_a = activeTag(a);
-                const tag_b = activeTag(b);
+                const tag_a: UnionTag = a;
+                const tag_b: UnionTag = b;
                 if (tag_a != tag_b) return false;
 
-                inline for (info.fields) |field_info| {
-                    if (@field(UnionTag, field_info.name) == tag_a) {
-                        return eql(@field(a, field_info.name), @field(b, field_info.name));
-                    }
-                }
-                return false;
+                return switch (a) {
+                    inline else => |val, tag| return eql(val, @field(b, @tagName(tag))),
+                };
             }
 
             @compileError("cannot compare untagged union type " ++ @typeName(T));
@@ -858,6 +855,15 @@ test eql {
 
     try testing.expect(eql(v1, v2));
     try testing.expect(!eql(v1, v3));
+
+    const CU = union(enum) {
+        a: void,
+        b: void,
+        c: comptime_int,
+    };
+
+    try testing.expect(eql(CU{ .a = {} }, .a));
+    try testing.expect(!eql(CU{ .a = {} }, .b));
 }
 
 test intToEnum {

--- a/lib/std/testing.zig
+++ b/lib/std/testing.zig
@@ -147,18 +147,10 @@ fn expectEqualInner(comptime T: type, expected: T, actual: T) !void {
 
             try expectEqual(expectedTag, actualTag);
 
-            // we only reach this loop if the tags are equal
-            inline for (std.meta.fields(@TypeOf(actual))) |fld| {
-                if (std.mem.eql(u8, fld.name, @tagName(actualTag))) {
-                    try expectEqual(@field(expected, fld.name), @field(actual, fld.name));
-                    return;
-                }
+            // we only reach this switch if the tags are equal
+            switch (expected) {
+                inline else => |val, tag| try expectEqual(val, @field(actual, @tagName(tag))),
             }
-
-            // we iterate over *all* union fields
-            // => we should never get here as the loop above is
-            //    including all possible values.
-            unreachable;
         },
 
         .Optional => {
@@ -206,6 +198,16 @@ test "expectEqual.union(enum)" {
     const a10 = T{ .a = 10 };
 
     try expectEqual(a10, a10);
+}
+
+test "expectEqual union with comptime-only field" {
+    const U = union(enum) {
+        a: void,
+        b: void,
+        c: comptime_int,
+    };
+
+    try expectEqual(U{ .a = {} }, .a);
 }
 
 /// This function is intended to be used only in tests. When the formatted result of the template
@@ -809,7 +811,7 @@ fn expectEqualDeepInner(comptime T: type, expected: T, actual: T) error{TestExpe
 
             try expectEqual(expectedTag, actualTag);
 
-            // we only reach this loop if the tags are equal
+            // we only reach this switch if the tags are equal
             switch (expected) {
                 inline else => |val, tag| {
                     try expectEqualDeep(val, @field(actual, @tagName(tag)));

